gemd/units/impl.py

```diff
@@ -3,70 +3,132 @@
 from pint.errors import DimensionalityError as IncompatibleUnitsError  # noqa Import
 from pint.errors import UndefinedUnitError, DefinitionSyntaxError  # noqa Import
 
 import functools
 import pkg_resources
-from typing import Union
+from typing import Union, List, Tuple
 
 # use the default unit registry for now
 DEFAULT_FILE = pkg_resources.resource_filename("gemd.units", "citrine_en.txt")
 _ALLOWED_OPERATORS = {".", "+", "-", "*", "/", "//", "^", "**", "(", ")"}
 
 
+def _space_after_minus_preprocessor(input_string: str) -> str:
+    """A preprocessor that protects against a pint < 0.21 bug."""
+    return re.sub(r"(?<=-)\s+(?=\d)", "", input_string)
+
+
+def _scientific_notation_preprocessor(input_string: str) -> str:
+    """Preprocessor that converts x * 10 ** y format to xEy."""
+    def _as_scientific(matchobj: re.Match) -> str:
+        return f"{matchobj.group(1) or '1'}e{matchobj.group(2)}"
+
+    number = r'\b(?:(\d+\.?\d*|\.\d+)\s*\*\s*)?10\s*(?:\*{2}|\^)\s*\+?(-?\d+\b)'
+    return re.sub(number, _as_scientific, input_string)
+
+
 def _scaling_preprocessor(input_string: str) -> str:
     """Preprocessor that turns scaling factors into non-dimensional units."""
+    blocks: List[List[Token]] = [[]]
+    operator_stack = []
+    for token in tokenizer(input_string):
+        exponent_context = any(t.string in {"**", "^"} for t in operator_stack)
+        if token.type == OP:
+            if token.string not in _ALLOWED_OPERATORS:
+                raise UndefinedUnitError(f"Unrecognized operator: {token.string}")
+
+            if exponent_context or token.string in {"**", "^", ".", "-", "+"}:
+                # Exponents & unaries do not change context
+                blocks[-1].append(token)
+            elif token.string not in {}:
+                blocks.append([])
+
+            if token.string == '(':
+                operator_stack.append(token)
+            elif token.string == ')':
+                while operator_stack:  # don't worry about enforcing balance
+                    if operator_stack.pop().string == '(':
+                        break  # We found token's friend
+            elif token.string in {"**", "^"}:
+                operator_stack.append(token)
+                continue  # Break flow since next token is in exponent context
+        elif token.type == NAME:
+            if exponent_context or len(blocks[-1]) == 0 or blocks[-1][-1].type != NAME:
+                blocks[-1].append(token)
+            else:  # Break blocks for two units in a row
+                blocks.append([token])
+        elif token.type == NUMBER:
+            blocks[-1].append(token)
+        elif token.type == ERRORTOKEN:  # Keep non-legal Python symbols like °
+            blocks[-1].append(token)
+        # Drop other tokens, such as EOF
+
+        if len(operator_stack) > 0 and operator_stack[-1].string in {"**", "^"}:
+            operator_stack.pop()  # Exit context for this exponential
```

Comment:
This breaks the complete expression into a series of terms that are separated by multiplication or division -- essentially the basic units we might want to substitute out.